File: project2/SowThistle.java
import java.util.Random;
import java.io.Serializable;

class SowThistle extends Plant implements Serializable {
    private static final long serialVersionUID = 1L;
    private static final Random random = new Random();
    private static final int WIDTH = 40;
    private static final int HEIGHT = 20;
    
    public SowThistle(int x, int y, World w) {
        super(0, x, y, w);
    }
    
    @Override
    public char draw() {
        return 's';
    }
    
    @Override
    public void action() {
        for (int i = 0; i < 3; i++) {
            if (random.nextInt(10) == 0) {
                int nx = x + (random.nextInt(3) - 1);
                int ny = y + (random.nextInt(3) - 1);
                if (nx >= 0 && ny >= 0 && nx < WIDTH && ny < HEIGHT && world.grid[ny][nx] == null) {
                    world.queueOrganism(new SowThistle(nx, ny, world));
                }
            }
        }
        age++;
    }
}

File: project3/sowthistle.py
from plant import Plant
import random

class SowThistle(Plant):
    def __init__(self, x, y, world):
        super().__init__(0, x, y, world)
    
    def action(self):
        # SowThistle attempts to spread 3 times per turn
        for _ in range(3):
            if random.randint(0, 9) == 0:  # 10% chance each attempt
                nx = self.x + (random.randint(0, 2) - 1)
                ny = self.y + (random.randint(0, 2) - 1)
                
                if (0 <= nx < self.WIDTH and 0 <= ny < self.HEIGHT and 
                    self.world.grid[ny][nx] is None):
                    # Create a new plant of the same type
                    new_plant = SowThistle(nx, ny, self.world)
                    self.world.queue_organism(new_plant)
        
        self.age += 1
    
    def draw(self):
        return 's'
